From 6a8eb3e5aef1045e77b8e069475232429ca974a3 Mon Sep 17 00:00:00 2001
From: Xiaowen Liu <b37945@freescale.com>
Date: Thu, 6 Dec 2012 09:36:34 +0800
Subject: [PATCH 382/635] ENGR00215174 add YU12 support for camera hal.

add YU12 support for camera hal.
fix cts verifier bugs.

Change-Id: I2cdd510cef584055767815bba3930e0a1afc688c
Signed-off-by: Xiaowen Liu <b37945@freescale.com>
---
 mx6/libcamera/CameraHal.cpp     | 10 +++++-----
 mx6/libcamera/DeviceAdapter.cpp |  2 +-
 mx6/libcamera/DeviceAdapter.h   |  4 ++--
 mx6/libcamera/Ov5640.cpp        |  4 ++++
 mx6/libcamera/Ov5642.cpp        |  4 ++++
 5 files changed, 16 insertions(+), 8 deletions(-)

diff --git a/mx6/libcamera/CameraHal.cpp b/mx6/libcamera/CameraHal.cpp
index f2dc61f..7bf8c33 100644
--- a/mx6/libcamera/CameraHal.cpp
+++ b/mx6/libcamera/CameraHal.cpp
@@ -118,27 +118,27 @@ void CameraHal::setCallbacks(camera_notify_callback         notify_cb,
 void CameraHal::enableMsgType(int32_t msgType)
 {
     if (mMsgEnabled & CAMERA_MSG_PREVIEW_FRAME) {
-        FLOGI("Enabling Preview Callback");
+        FLOG_RUNTIME("Enabling Preview Callback");
     }
     else {
-        FLOGI("Preview callback not enabled %x", msgType);
+        FLOG_RUNTIME("Preview callback not enabled %x", msgType);
     }
 
     Mutex::Autolock lock(mLock);
     mCameraBridge->enableMsgType(msgType);
-    FLOGI("enableMsgType 0x%x", msgType);
+    FLOG_RUNTIME("enableMsgType 0x%x", msgType);
     mMsgEnabled |= msgType;
 }
 
 void CameraHal::disableMsgType(int32_t msgType)
 {
     if (msgType & CAMERA_MSG_PREVIEW_FRAME) {
-        FLOGI("Disabling Preview Callback");
+        FLOG_RUNTIME("Disabling Preview Callback");
     }
 
     Mutex::Autolock lock(mLock);
     mCameraBridge->disableMsgType(msgType);
-    FLOGI("disableMsgType 0x%x", msgType);
+    FLOG_RUNTIME("disableMsgType 0x%x", msgType);
     mMsgEnabled &= ~msgType;
 }
 
diff --git a/mx6/libcamera/DeviceAdapter.cpp b/mx6/libcamera/DeviceAdapter.cpp
index d5ef601..5a13dbd 100644
--- a/mx6/libcamera/DeviceAdapter.cpp
+++ b/mx6/libcamera/DeviceAdapter.cpp
@@ -527,7 +527,7 @@ int DeviceAdapter::deviceThread()
     }
 
     dispatchCameraFrame(frame);
-    if (mImageCapture) {
+    if (mImageCapture || !mPreviewing) {
         FLOGI("device thread exit after take picture");
         return ALREADY_EXISTS;
     }
diff --git a/mx6/libcamera/DeviceAdapter.h b/mx6/libcamera/DeviceAdapter.h
index ae52542..cb4a295 100644
--- a/mx6/libcamera/DeviceAdapter.h
+++ b/mx6/libcamera/DeviceAdapter.h
@@ -43,11 +43,11 @@ public:
                                      PixelFormat format,
                                      int         fps);
     PixelFormat getPreviewPixelFormat() {
-        return mPicturePixelFormat;
+        return mPreviewPixelFormat;
     }
 
     PixelFormat getPicturePixelFormat() {
-        return mPreviewPixelFormat;
+        return mPicturePixelFormat;
     }
 
     virtual status_t setParameters(CameraParameters& params) = 0;
diff --git a/mx6/libcamera/Ov5640.cpp b/mx6/libcamera/Ov5640.cpp
index 969cedd..4403e53 100644
--- a/mx6/libcamera/Ov5640.cpp
+++ b/mx6/libcamera/Ov5640.cpp
@@ -201,6 +201,10 @@ status_t Ov5640::setParameters(CameraParameters& params)
               params.getPreviewFormat());
         return BAD_VALUE;
     }
+    else {
+        mPreviewPixelFormat = convertStringToPixelFormat(
+                                 params.getPreviewFormat());
+    }
 
     if (strcmp(params.getPictureFormat(), "jpeg") != 0) {
         FLOGE("Only jpeg still pictures are supported");
diff --git a/mx6/libcamera/Ov5642.cpp b/mx6/libcamera/Ov5642.cpp
index 610bed1..73aead8 100644
--- a/mx6/libcamera/Ov5642.cpp
+++ b/mx6/libcamera/Ov5642.cpp
@@ -201,6 +201,10 @@ status_t Ov5642::setParameters(CameraParameters& params)
               params.getPreviewFormat());
         return BAD_VALUE;
     }
+    else {
+        mPreviewPixelFormat = convertStringToPixelFormat(
+                                 params.getPreviewFormat());
+    }
 
     if (strcmp(params.getPictureFormat(), "jpeg") != 0) {
         FLOGE("Only jpeg still pictures are supported");
-- 
1.8.0

